[WKINT-405] Add agent + APM installation powershell module

ISauve opened the pull request:
<!--
* New contributors are highly encouraged to read our
  [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* Both Contributor and Reviewer Checklists are available at https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#pull-requests.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.
-->
### What does this PR do?

<!--
* A brief description of the change being made with this pull request.
* If the description here cannot be expressed in a succinct form, consider
  opening multiple pull requests instead of a single one.
-->

### Motivation

<!--
* What inspired you to submit this pull request?
* Link any related GitHub issues or PRs here.
-->

### Additional Notes

<!--
* Anything else we should know when reviewing?
* Include benchmarking information here whenever possible.
* Include info about alternatives that were considered and why the proposed
  version was chosen.
-->

### Possible Drawbacks / Trade-offs

<!--
* What are the possible side-effects or negative impacts of the code change?
-->

### Describe how to test/QA your changes

<!--
* Write here in detail or link to detailed instructions on how this change can
  be tested/QAd/validated, including any environment setup.
-->
Changes have automated tests, but this is how to do it manually:

On a fresh VM:

1. Import the module into your current powershell session
```
> Import-Module <path/to>\Datadog
```
2. Verify that only the `Install-DDAgent` function has been imported
```
> Get-Command -Module DDInstall

CommandType     Name                                               Version    Source
-----------     ----                                               -------    ------
Function        Install-DDAgent                                    1.0        datadog
```
3. Print the documentation for the command & validate that the details are all correct
```
 > Get-Help Install-DDAgent
 ```
 4. Install an old version of the agent with the dotnet tracer, validating that tabbing to auto-complete works for the function, the parameter names, and the tracer name (dotnet)
 ```
 >  Install-DDAgent -WithAPMTracers DotNet -AgentInstallURL "https://s3.amazonaws.com/ddagent-windows-stable/datadog-agent-7.48.0-1.x86_64.msi"
 ```
 5. Validate that both installs succeeded
 6. Upgrade to the newest agent
 ```
 >  Install-DDAgent
 ```
 7. Validate that the upgrade succeeded
 8. Test some error conditions (providing an invalid install URL or invalid tracing library, downgrading to an earlier agent version, running from a non-administrator prompt, etc.)

ISauve committed:
Add Windows agent + APM installation powershell module

ISauve committed:
Flesh out module

ISauve committed:
Move the function comments so Get-Help can find them

ISauve committed:
Fix install error check

ISauve commented on DDInstall/DDInstall.psm1:
```diff
@@ -0,0 +1,192 @@
+function Install-DDAgent
```
Since Github doesn't seem to know how to display `psd1` files, I'm copying the important parts here for visibility:
```
# Script module or binary module file associated with this manifest.
RootModule = 'DDInstall.psm1'

# Version number of this module.
ModuleVersion = '1.0'

# Supported PSEditions
# CompatiblePSEditions = @()

# ID used to uniquely identify this module
GUID = 'd8ed3d7e-ed2f-4e84-9987-02fcfa3aed19'

# Company or vendor of this module
CompanyName = 'Datadog'

# Copyright statement for this module
Copyright = '(c) 2024 Datadog. All rights reserved.'

# Functions to export from this module, for best performance, do not use wildcards and do not delete the entry, use an empty array if there are no functions to export.
FunctionsToExport = @("Install-DDAgent")
```

One part I think is important to discuss is the module's name - I used `DDInstall` as a placeholder, but I don't love it. @richl-dd, any suggestions?

richl-dd replied:
How does everyone feel about just calling it Datadog? I looked through the most popular packages (https://www.powershellgallery.com/packages) and the convention seems to be either a single name (e.g. `Datadog`) or something like `CompanyName.Thing.SubThing` (e.g. `Microsoft.WinGet.Client`).

I think `Install-Module Datadog` would be fine - thoughts (@shanelhuang @ISauve @clarkb7 @derekwbrown)?

ISauve replied:
Works for me!

ISauve commented on DDInstall/DDInstall.psm1:
```diff
@@ -0,0 +1,192 @@
+    param (
+        [Parameter()][ValidateSet('DotNet')][String[]]$WithAPMTracers,
+        [String] $AgentInstallURL,
+        [Parameter(Mandatory)] [String]$ApiKey,
+        [String]$Site,
+        [String]$Tags,
+        [String]$Hostname,
+        [String]$DDAgentUsername,
+        [String]$DDAgentPassword,
+        [String]$ApplicationDataDirectory,
+        [String]$ProjectLocation,
```
This list of parameters comes from [our installer docs](https://docs.datadoghq.com/agent/basic_agent_usage/windows/?tab=commandline#installation-configuration-options).

There are other options we could choose to support (ie [these ones](https://docs.datadoghq.com/agent/basic_agent_usage/windows/?tab=commandline#more-agent-configuration-options)). We could also add these parameters explicitly, which has a couple advantages (ie properly documenting the supported options + adding support for auto-complete), or we could add a "catch all" parameter which would accept any number of additional options in the form `KEY=VALUE`. I'm open to either approach (or something different entirely!)

shanelhuang replied:
I don't think supporting the additional installer options in the second link is necessary. Ideally customers should enable those features in the datadog.yaml file

clarkb7 replied:
The script could check if `datadog.yaml` already exists and throw display a warning if the options are provided. A terminating error might be a bit much since then initial install and future installs would need different command lines.
`ApplicationDataDIrectory` is provided here and could be used for the path.
It's also in the registry at `HKLM\SOFTWARE\Datadog\Datadog Agent\ConfigRoot`

derekwbrown replied:
> I don't think supporting the additional installer options in the second link is necessary. Ideally customers should enable those features in the datadog.yaml file

I disagree, at least for customers doing a first time install.  The reason we have the commadn line options is because editing the yaml for basic first time install is non-trivial, so we put the most basic of options on the command line.

Also the agent installer will ignore most, but  not all, of these options on upgrade.

I would suggest the following:
1) if it's a new install, or a reinstall but `datadog.yaml` doesn't have an api key, process everything as you have above (this is what the installer does)
2) if it's an upgrade/there's already a working config, don't pass _any_ options.  If the user passes any options, then either (a) put up yellow warning that they'll be ignored and continue or (b) fail the script and make them start it with no options.

either way with (2) they should be aware those options won't be accepted.

Suggest coordinating with #windows-agent to figure out which options are actually honored when, and as above either fail when they're going to be ignored, or at least print a warning.

richl-dd replied:
Derek, you'd have no way of knowing this, but in a separate Slack discussion while you were OOO, we discussed eventually wanting to remove these options from being accessible by the installer at all (i.e. installer editing the YAML) because customers find the behavior exceptionally confusing as to when the options are respected or not.

We'd prefer to steer customers away from passing them to the installer and towards using RC or your idea of an on-host GUI configurator. If we wanted them available via this module, then I'd suggest not passing them to the install (for reasons mentioned above) and instead have the module enable them directly (e.g. separate module commands that edit the YAML as opposed to the MSI installer).

All that to say, I think that any support for these options is outside the scope of the first version of the module, which is geared towards APM SSI. We can discuss if/how we want them supported in a later version.

ISauve commented on DDInstall/DDInstall.psm1:
```diff
@@ -0,0 +1,192 @@
+function Install-DDAgent
+{
+    param (
+        [Parameter()][ValidateSet('DotNet')][String[]]$WithAPMTracers,
```
note that I only started with support for .NET, but we can add other tracing libraries by simply adding other values to this set

ISauve replied:
note that I only started with support for .NET, but we can add other tracing libraries by simply adding other values to this set

clarkb7 commented on DDInstall/DDInstall.psm1:
```diff
@@ -0,0 +1,192 @@
+{
+    $ProgressPreference = 'SilentlyContinue'
+
+    Invoke-WebRequest -Uri $url -OutFile $outFile
```
add [-UseBasicParsing](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.utility/invoke-webrequest?view=powershell-5.1#-usebasicparsing)

clarkb7 commented on DDInstall/DDInstall.psm1:
```diff
@@ -0,0 +1,192 @@
+    $ProgressPreference = 'SilentlyContinue'
+
+    Invoke-WebRequest -Uri $url -OutFile $outFile
+    # Suppress progress bar to prevent performance issues (affects older windows versions, see https://github.com/PowerShell/PowerShell/issues/13414)
```
should this comment be up a few lines? Also I guess this means there wasn't a way to detect if we were running interactively or not :/

ISauve replied:
ah yep, copy-paste error 😬

I found a couple recommendations of ways to check if the session was interactive, but I wasn't sure how reliable they were & figured it might just be safer to always disable it. But, if you think that showing a download progress bar for interactive sessions is worth doing, I can definitely add it & just spend some time testing it to make sure it's reliable!

richl-dd replied:
>  if you think that showing a download progress bar for interactive sessions is worth doing
I feel pretty strong that it's worth doing - when we were debugging this with a customer, it took a _long_ time for everything to complete and we were left wondering if the script hung or not.

Do you know what the performance impact is?

clarkb7 commented on DDInstall/DDInstall.psm1:
```diff
@@ -0,0 +1,192 @@
+    $installResult = Start-Process -Wait msiexec -ArgumentList "/qn /i datadog-agent.msi $installerParameters" -PassThru
+    if ($installResult.ExitCode -ne 0)
+    {
+        $logFile = "%TEMP%\MSI*.LOG"
```
instead of not knowing what the exact log path if the user didn't set one, we could create our own default using [New-TemporaryFile](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.utility/new-temporaryfile?view=powershell-7.4), which would let us provide the exact path to the user in the case of an error.

ISauve replied:
another great idea!

clarkb7 commented on DDInstall/DDInstall.psm1:
```diff
@@ -0,0 +1,192 @@
+{
+    param (
+        [Parameter()][ValidateSet('DotNet')][String[]]$WithAPMTracers,
+        [String] $AgentInstallURL,
```
We should support a file path, too. in case an admin stores the MSI on a file share, or is running this module as part of a bundle, like from SCCM, and the MSI file is included alongside it.

ISauve replied:
good thinking!

clarkb7 commented on DDInstall/DDInstall.psm1:
```diff
@@ -0,0 +1,192 @@
+    Override the default ddagentuser username used during Agent installation (v6.11.0+).
+
+    .PARAMETER DDAgentPassword
+    Override the cryptographically secure password generated for the ddagentuser user during Agent installation (v6.11.0+). Must be provided for installs on domain servers.
```
s/domain servers/domain controllers

clarkb7 reviewed (commented):
very much looking forward to this!

ISauve committed:
Add AgentInstallerPath parameter

ISauve committed:
Create our own temporary log file by default

ISauve committed:
Fix formatting

ISauve committed:
Fix description of DDAgentPassword parameter

ISauve committed:
Add -UseBasicParsing and move comment to the correct location

clarkb7 commented on DDInstall/DDInstall.psm1:
```diff
@@ -0,0 +1,192 @@
+
+function downloadAsset($url, $outFile)
+{
+    $ProgressPreference = 'SilentlyContinue'
```
what is the scope of this option? Is it only for this function, or does it affect the rest of the console session after this point?

ISauve replied:
resolving since this was replaced with `(New-Object System.Net.WebClient).DownloadFile($url, $outFile)`

derekwbrown commented on DDInstall/DDInstall.psm1:
```diff
@@ -0,0 +1,215 @@
+    # Suppress progress bar to prevent performance issues (affects older windows versions, see https://github.com/PowerShell/PowerShell/issues/13414)
+    $ProgressPreference = 'SilentlyContinue'
+
+    Invoke-WebRequest -UseBasicParsing -Uri $url -OutFile $outFile
```
You could try
`(New-Object System.Net.WebClient).DownloadFile($url, $outFile)`

It gets rid of the progress banner (so you might be able to get rid of the `SilentlyContinue`.  It's also way faster.

derekwbrown replied:
Also, with this I believe you want to set a fully qualified target path ($outFile).

richl-dd replied:
^ is a good call - in a previous comment, I said I felt strongly about the progress bar because otherwise it took > 5 minutes of looking at what appeared to be a hung prompt. If we want to do away with the progress bar, then I think at a minimum we need to print "Download In Progress" or something, so we know that it is actually doing something.

I still feel strongly that we need some indication that something is happening / the script isn't broken, but I'm open to it not being a progress bar. Maybe like a `[=]` that spins so like: `[=]`  --> `[||]` -> `[=]`.

richl-dd replied:
(Yes, that looks horrible, but I don't know how else to represent it in a static comment 😅 )

derekwbrown replied:
since `DownloadFile` is a blocking call, I think you're stuck w/o active progress.
  
  You could certainly add `Write_host` checkpoints along the way, i.e.
  ```
  Beginning download file ...
  ... done download
  Beginning install...
  ... done install```

ISauve replied:
This is the current output:
```
> Install-DDAgent -WithAPMTracers DotNet -ApiKey xxxxx
Downloading Datadog Windows Agent installer
Installing Datadog Windows Agent
Downloading Datadog .NET Tracing Library v2.49.0
Installing .NET Tracing Library
```
Do we think that's sufficient?

derekwbrown commented on DDInstall/DDInstall.psm1:
```diff
@@ -0,0 +1,215 @@
+        return $false
+    }
+
+    if ($signature.SignerCertificate.Issuer -ne 'CN=DigiCert Trusted G4 Code Signing RSA4096 SHA384 2021 CA1, O="DigiCert, Inc.", C=US')
```
Suggest being a little less restrictive on the issuer; concern is that when we get a new signing cert, the issuer could change in subtle ways (I'm thinking the year as an obvious one), which would cause the script to just start failing.

richl-dd replied:
The concern is valid - I think we can mitigate it through manual and automated tests that will identify a mismatch. Since we're using this for validation that our module hasn't been tampered with, the more restrictive the better.

We should ensure we have automated tests for this AND this install method should always be included during manual QA. I think those 2 things are enough to enable us to keep this restrictive.

@ISauve is this something that we can add into automated tests?

@clarkb7 how much of a pain would including this into every release QA cycle be?

@derekwbrown if you still feel we need to loosen the restriction, what would you suggest?

derekwbrown replied:
Including in every cycle doesn't help.  Someone installs the module today.  Tomorrow we get a new cert that's slightly different.  then they run the command.  breaks for no reason.

So, to answer your questoin, I can think of a bunch of ways.
1) Least restrictive... as long as the signature is valid and has a valid chain, go for it.
2) More restrictive than (1), but more work.  Parse the CN of the cert, and make sure it includes (case insensitive) datadog.  Parse the CN of the issuer, make sure it  includes DigiCert.  Make sure the Object (O) includes digicert.

Even restricting to `DigiCert` is at least potentially flakey, if we ever change what CA we buy our certs from.

The OS is supposed to manage the trusted roots; tbh I think a valid signature is sufficient.

derekwbrown commented on DDInstall/DDInstall.psm1:
```diff
@@ -0,0 +1,215 @@
+        return $false
+    }
+
+    if ($signature.SignerCertificate.Subject -ne 'CN="Datadog, Inc", O="Datadog, Inc", L=New York, S=New York, C=US')
```
Not sure how feasible this is, but same comment here.  If the next code signing cert someone puts "Datadog, Inc." in the CSR, then I think this will break.

derekwbrown commented on DDInstall/DDInstall.psm1:
```diff
@@ -0,0 +1,215 @@
+    [System.Net.ServicePointManager]::SecurityProtocol = [System.Net.ServicePointManager]::SecurityProtocol -bor [System.Net.SecurityProtocolType]::Tls12
+
+    $installerPath = "datadog-agent.msi"
+    if ($PSBoundParameters.ContainsKey('AgentInstallerPath'))
```
For this (and for the tracer msi below), I think you want to create a filename in the temporary directory, so that you have a fully qualified name.  
a) the suggested download below requires a fq path
b) you can (and should) clean up the downloaded MSI when done.  as written, I think this will leave 1 or 2 MSIs around in the current working directory, whereever that might be.

richl-dd replied:
This is a great catch, thank you

ISauve replied:
great idea!

ISauve committed:
Download MSIs to temporary directory

ISauve committed:
Use 'DownloadFile' instead of Invoke-WebRequest

ISauve committed:
Rename module

ISauve committed:
Delete installers after execution

ISauve committed:
Print warning if datadog.yaml file already exists + add support for downloading specific agent version

ISauve committed:
Move agent & tracer installs into helper functions

ISauve committed:
Reduce restrictiveness of signature check

ISauve committed:
Wrap registry key check in try/catch

ISauve committed:
Fix catch for ItemNotFound exception in doesDatadogYamlExist

ISauve committed:
Let installer determine whether or not to ignore parameters

ISauve committed:
Pass installer arguments as a list

Fixes problems with passing values that are surrounded by quotes

ISauve committed:
Add UseBasicParsing to .NET tracer download

ISauve committed:
Output error in x86 shells

ISauve committed:
Add E2E test for powershell module

ISauve committed:
Use testClient.GetAgentVersion() everywhere instead of helper

derekwbrown committed:
write install_info file

CelianR reviewed (approved):
LGTM for gitlab related files

derekwbrown committed:
bring tests into the present

derekwbrown committed:
add restart iis command if dotnet tracer installed

derekwbrown committed:
Update the command with new arguments

Update help text to match updated commands

Update tests to match updated arguments

derekwbrown committed:
Final changes to argument parsing.

Also add `preview` flag which allows user to see what would happen prior
to executing (do a dry run).

derekwbrown committed:
Put back checking for admin even in preview mode; it should fail in
preview mode the same as a real run.

derekwbrown committed:
fix function name in module definition file

derekwbrown commented:
/merge
